rllib/offline/estimators/qreg_torch_model.py

```diff
@@ -0,0 +1,160 @@
+from ray.rllib.models.utils import get_initializer
+from ray.rllib.policy import Policy
+from typing import Dict, List, Union
+
+from ray.rllib.models.catalog import ModelCatalog
+from ray.rllib.models.torch.torch_modelv2 import TorchModelV2
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.framework import try_import_torch
+from ray.rllib.utils.typing import TensorType
+
+torch, nn = try_import_torch()
+
+
+class QRegTorchModel(TorchModelV2):
+    """Pytorch implementation of the Q-Reg model from
+    https://arxiv.org/pdf/1911.06854.pdf
+
+    Arguments:
+        policy: The Policy object correspodning to the target policy in OPE
+        gamma: The discount factor for the environment
+        config: Optional config settings for Q-Reg
+        config = {
+            # The ModelConfigDict for self.q_model
+            "model": {"fcnet_hiddens": [32, 32], "fcnet_activation": "relu"},
+            # Maximum number of training iterations to run on the batch
+            "n_iters": 80,
+            Learning rate for Q-function optimizer
+            "lr": 1e-3,
+            Early stopping if the mean loss < delta
+            "delta": 1e-4,
+        }
+    """
+
+    def __init__(self, policy: Policy, gamma: float, config: Dict) -> None:
+        self.policy = policy
+        self.gamma = gamma
+        self.observation_space = policy.observation_space
+        self.action_space = policy.action_space
+
+        self.q_model: TorchModelV2 = ModelCatalog.get_model_v2(
+            self.observation_space,
+            self.action_space,
+            self.action_space.n,
+            config.get("model", {}),
+            framework="torch",
+            name="TorchQModel",
+        )
+        self.device = self.policy.device
+        self.n_iters = config.get("n_iters", 80)
+        self.lr = config.get("lr", 1e-3)
+        self.delta = config.get("delta", 1e-4)
+        self.optimizer = torch.optim.Adam(self.q_model.variables(), self.lr)
+        initializer = get_initializer("xavier_uniform", framework="torch")
+
+        def f(m):
+            if isinstance(m, nn.Linear):
+                initializer(m.weight)
+
+        self.initializer = f
+
+    def reset(self) -> None:
+        """Resets/Reinintializes the model weights."""
+        self.q_model.apply(self.initializer)
+
+    def train_q(self, batch: SampleBatch) -> TensorType:
```

Comment:
I don't know if it's too much to ask. what do you think if we add some unit tests for this model? especially the train() impl?

Reply:
@rapotdar and I were discussing offline to add "learning" tests for (at least the new) methods. Something like:
- Learn CartPole with some offline data for some time.
- Estimate the learnt policy.
- Check, whether estimation is correct by comparing it with the actual performance of the learnt policy in an actual env.

One big problem with this is variance. But I think it would be really realy cool to have these tests, even if the fault tolerance has to be large. Would also allow us to compare different OPE methods with each other.

Reply:
We can do this in a follow up PR, I think to not make this one blow up. :)